pkg/agent/manager/cache/cache.go

```diff
@@ -0,0 +1,182 @@
+	DeleteEntry(regEntry *common.RegistrationEntry) bool
+	// Entries returns all the in force cached entries.
+	Entries() chan Entry
+	// IsEmpty returns true if this cache doesn't have any entry.
+	IsEmpty() bool
+	// Register a Subscriber and return WorkloadUpdate on the subscriber's channel
+	Subscribe(sub *Subscriber)
+}
+
+type cacheImpl struct {
+	// Map keyed by a combination of SpiffeId + ParentId + Selectors holding a list of
+	// Entry instances ordered by SVID expiration date.
+	cache       map[string][]Entry
+	log         logrus.FieldLogger
+	m           sync.Mutex
+	Subscribers *subscribers
+	bundle      []*x509.Certificate
+}
+
+// New creates a new Cache.
+func New(log logrus.FieldLogger, bundle []*x509.Certificate) Cache {
+	return &cacheImpl{
+		cache:       make(map[string][]Entry),
+		log:         log.WithField("subsystem_name", "cache"),
+		bundle:      bundle,
+		Subscribers: NewSubscribers(),
+	}
+}
+
+func (c *cacheImpl) SetServerBundle(bundle []*x509.Certificate) {
+	c.m.Lock()
+	defer c.m.Unlock()
+	c.bundle = bundle
+}
+
+func (c *cacheImpl) serverBundle() []*x509.Certificate {
+	c.m.Lock()
+	defer c.m.Unlock()
+	return c.bundle
+}
+
+func (c *cacheImpl) Entries() chan Entry {
+	c.m.Lock()
+	defer c.m.Unlock()
+	entries := make(chan Entry, len(c.cache))
+	for _, e := range c.cache {
+		// Only return the first element for each array of entries because it is the
+		// in force entry.
+		entries <- e[0]
+	}
+	close(entries)
+	return entries
+}
+
+func (c *cacheImpl) Subscribe(sub *Subscriber) {
+	entries := c.Entries()
+	c.m.Lock()
+	defer c.m.Unlock()
+	subEntries := SubscriberEntries(sub, entries)
+	go func() {
+		select {
+		case <-sub.done:
+			c.Subscribers.remove(sub)
+			close(sub.C)
+		case sub.C <- &WorkloadUpdate{Entries: subEntries, Bundle: c.bundle}:
+		}
+	}()
+
+	c.Subscribers.Add(sub)
+}
+
+func (c *cacheImpl) Entry(regEntry *common.RegistrationEntry) *Entry {
+	key := util.DeriveRegEntryhash(regEntry)
+	c.m.Lock()
+	defer c.m.Unlock()
+	if entries, found := c.cache[key]; found {
+		return &entries[0]
+	}
+	return nil
+}
+
+func (c *cacheImpl) SetEntry(entry *Entry) {
+	entries := c.Entries()
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	key := util.DeriveRegEntryhash(entry.RegistrationEntry)
+	c.cache[key] = append(c.cache[key], *entry)
+
+	subs := c.Subscribers.Get(entry.RegistrationEntry.Selectors)
+	for _, sub := range subs {
+		subEntries := SubscriberEntries(sub, entries)
+		select {
+		case <-sub.done:
+			c.Subscribers.remove(sub)
+			close(sub.C)
+		case sub.C <- &WorkloadUpdate{Entries: subEntries, Bundle: c.bundle}:
+		}
+	}
+	return
+}
+
+func (c *cacheImpl) DeleteEntries(regEntry *common.RegistrationEntry) int {
+	c.m.Lock()
+	defer c.m.Unlock()
+	key := util.DeriveRegEntryhash(regEntry)
+	if entries, found := c.cache[key]; found {
+		delete(c.cache, key)
+		return len(entries)
+	}
+	return 0
+}
+
+func (c *cacheImpl) DeleteEntry(regEntry *common.RegistrationEntry) bool {
```

Comment:
I think we need to send subscriber notifications for workloads which were issued the entry being deleted